3rdParty/iresearch/core/formats/columnstore2.cpp

```diff
@@ -801,82 +909,176 @@ doc_iterator::ptr fixed_length_column::iterator(bool /*consolidation*/) const {
+        buffered_input_ = std::make_unique<irs::remapped_bytes_ref_input>(
+            irs::bytes_ref{column_data_.data(), column_data_.size()},
+            std::move(mapping));
+      }
+    }
+    if (buffered_input_) {
+      reset_stream(buffered_input_.get());
+    }
+  }
+
  private:
+  static std::vector<column_block> read_blocks_sparse(const column_header& hdr,
+                                                      index_input& in);
+
   template<typename ValueReader>
   class payload_reader : private ValueReader {
    public:
     template<typename... Args>
-    payload_reader(
-        const column_block* blocks,
-        Args&&... args)
-      : ValueReader{ std::forward<Args>(args)... },
-        blocks_{blocks} {
-    }
+    payload_reader(const column_block* blocks, Args&&... args)
+        : ValueReader{std::forward<Args>(args)...}, blocks_{blocks} {}
 
     bytes_ref payload(doc_id_t i);
 
    private:
     const column_block* blocks_;
-  }; // payload_reader
+  };  // payload_reader
+
+  template<bool encrypted>
+  bool make_buffered_data(
+      column_header& hdr, irs::index_input& in,
+      std::vector<column_block>& blocks,
+      std::vector<irs::byte_type>& column_data,
+      range<std::unique_ptr<column_reader>> next_sorted_columns,
+      remapped_bytes_ref_input::mapping* mapping) {
+    // idx adr/block offset length source
+    std::vector <std::tuple<size_t, bool, size_t, size_t, size_t>> chunks;
+    size_t chunks_size{0};
+    size_t block_idx{0};
+    std::vector<irs::byte_type> addr_buffer;
+    for (auto& block : blocks) {
+      if (bitpack::ALL_EQUAL == block.bits) {
+        size_t length = block.avg * block.last;
+        length += block.last_size;
+        chunks.emplace_back(block_idx++, false, chunks_size, length, block.data);
+        chunks_size += length;
+      } else {
+        // addr table size
+        const size_t addr_length = packed::bytes_required_64(
+            math::ceil64((block.last + 1), packed::BLOCK_SIZE_64), block.bits);
+        chunks.emplace_back(block_idx, true, chunks_size, addr_length, block.addr);
+        chunks_size += addr_length;
+        // block data size is calculated as end of the data for the last document in block
+        const size_t block_size = block.bits * sizeof(uint64_t);
+        auto const value_index = block.last % packed::BLOCK_SIZE_64;
+        addr_buffer.resize(block_size);
+        in.read_bytes(block.addr + addr_length - block_size,
+                       addr_buffer.data(), block_size);
+        const uint64_t start_delta = zig_zag_decode64(packed::fastpack_at(
+            reinterpret_cast<const uint64_t*>(addr_buffer.data()),
+            value_index, block.bits));
+        const uint64_t start = block.avg * block.last + start_delta;
+
+        size_t length = block.last_size + start;
+        chunks.emplace_back(block_idx++, false, chunks_size, length, block.data);
+        chunks_size += length;
```

Comment:
These two lines are the same for the `if` and `else` cases, and could be executed unconditionally.